### PR TITLE
Define the record_stream method in native_functions.yaml

### LINKED ISSUE #36556: Move the record_stream to Aten operation
## 🚀 Feature
Decouple the Stream and Event with the CUDA.

## Motivation
While the stream based synchronization mechanism are widely used and generalized in the c10, c10 stream to cuda stream, the python code still coupled with cuda stream, only torch.cuda.stream and torch.cuda.event are available.

Need to decouple the python stream and event to support more hardware other than the cuda devices.

## Pitch
1. Add a python class to expose the c10 Stream and the torch cuda stream should inherit from the common stream.
2. Add a new class to abstract the CUDAEvent and the torch cuda event should inherit from the common base event.
3. Generalize the specific CUDA code for common abstraction: like move the hard coded THPVariable_record_stream to the ATen operation.

### PR DESCRIPTION
**Pull request by chengjunlu**
The record_stream method was hard coded for CUDA device. Define the record_stream in the native_functions.yaml to enable the dynamic dispatch to different end device.

Fixes #36556

### CODE REVIEW

**Comment by chengjunlu** [comment]
@mrshenli Would you like to help review this PR. 
This change is to decouple the tensor record_stream method to CUDA.

**Comment by ezyang** [comment]
This is a really clean PR, thank you for the work.

**Comment by ezyang** [comment]
Most of this all looks good. The part I am most nervous about is the Python binding code.

File: torch/csrc/Stream.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
+#include <pybind11/pybind11.h>
+#include <torch/csrc/Device.h>
+#include <torch/csrc/THP.h>
+#include <torch/csrc/utils/python_arg_parser.h>
```
Comment (ezyang) [inline torch/csrc/Stream.cpp:4]: I'm guessing this was modeled off of cuda/Stream.cpp

File: torch/csrc/Stream.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  {nullptr}
+};
+
+static PyMethodDef THPStream_methods[] = {
+  {(char*)"__eq__", (PyCFunction)THPStream_eq, METH_O, nullptr},
+  {nullptr}
+};
+
+PyTypeObject THPStreamType = {
+  PyVarObject_HEAD_INIT(&PyType_Type, 0)
```
Comment (ezyang) [inline torch/csrc/Stream.cpp:60]: Huh, THCPStreamType previously was nullptr here. Is this the right way to do it?

**Comment by ezyang** [comment]
XLA is borken:

```
Sep 17 04:15:41 In file included from torch_xla/csrc/aten_xla_type.cpp:15:
Sep 17 04:15:41 /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type_default.h:757:48: error: unknown type name 'Stream'
Sep 17 04:15:41   static void record_stream(at::Tensor & self, Stream s);
Sep 17 04:15:41                                             
```

If Stream gets translated into a fully qualified type that probably would fix it.

File: torch/csrc/Stream.cpp (hunk 2 above)
Comment (chengjunlu) [reply torch/csrc/Stream.cpp:60]: PyType_Type introduce more properties for CLASS type. Both nullptr and PyType_Type work. Right?

File: torch/csrc/Stream.cpp (hunk 1 above)
Comment (chengjunlu) [reply torch/csrc/Stream.cpp:4]: Yes. A common base Stream for all the devices.

**Comment by chengjunlu** [comment]
Thank you for your effort on reviewing this PR and your encouragements.

**Comment by ezyang** [comment]
Any progress on the XLA problem? Do you need help with it?

**Comment by chengjunlu** [comment]
> Any progress on the XLA problem? Do you need help with it?

@ezyang XLA PR [2542](https://github.com/pytorch/xla/pull/2542) fixes the issue.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
I'm attempting to land this. To tree hugging oncall: if this PR is causing XLA to fail, please don't revert; we just need to land the companion XLA PR.

File: aten/src/ATen/core/jit_type.h [hunk 3]
```diff
@@ -1550,6 +1551,28 @@ struct CAFFE2_API DeviceObjType : public Type {
   DeviceObjType() : Type(TypeKind::DeviceObjType) {}
 };
 
+struct StreamObjType;
+using StreamObjTypePtr = std::shared_ptr<StreamObjType>;
+// This type represents a Generator
```
Comment (eellison) [inline aten/src/ATen/core/jit_type.h:1556]: This is the same comment as from GeneratorType - copy pasta ?

Comment (chengjunlu) [reply aten/src/ATen/core/jit_type.h:1556]: Yes. This comment is incorrectly.

**Pull request closed without merging**